### PR TITLE
Special array index to replace b[...] with *b

### PR DESCRIPTION
**Pull request by tahina-pro**
Access to a F* (or Steel) reference `r` is currently extracted to C as `r[0]`, both for reading and writing. Currently, the only way to produce `*r` instead has been to use the special operators in `LowStar.BufferOps`, but this only works with Low* buffers.

**This PR** proposes an alternative way: I introduce a new constant, `C._zero_for_deref`, defined as 0, and I change C code generation from `r[C._zero_for_deref]` to `*r`. This avoids adding a new node in the Karamel source AST.

I have an Everest green (project-everest/everest@35b47bb1ac32eb01f6f507dc65d00ebe0e5609aa) with a companion F* commit (FStarLang/FStar@984a531b50b07794d4edd91bc47ccaaa13546796) that enacts such code generation improvement for F* and Steel references, for which I'll open a F* PR shortly.

### CODE REVIEW

**Comment by msprotz** [comment]
Looks good. I have only one remark: you might find it hard to eliminate code-generation for this specific value, since the dependency analysis (in the current state of your patch) is not aware that there is a special treatment for it down the line.

Have you thought about that?

**Comment by tahina-pro** [comment]
> Looks good. I have only one remark: you might find it hard to eliminate code-generation for this specific value, since the dependency analysis (in the current state of your patch) is not aware that there is a special treatment for it down the line.

Thanks for pointing this out. Yes, I do observe that in my EverParse experiments: a definition of `_zero_for_deref` is inserted somewhere in my bundles. (This also seems to explain why `_zero_for_deref` appears in the generated `dist/generic/C.h`)
I am not aware of any mechanism within Karamel to declare a list of symbols with special treatment for which no C definition should be produced. If there is no such mechanism, then the only way I can see could be to add an attribute to be conveyed down to Karamel and not produce any C code for definitions tagged with that attribute.
Ideally, I would have reused `noextract_to "krml"`, but then the definition already disappears from the `.krml` file and Karamel complains when it is used. That's why we seem to need another separate attribute.
What do you think?

**Comment by msprotz** [comment]
I don't think you need a new attribute.

**Comment by msprotz** [comment]
Thanks, this looks good. I'm not sure your test does very much in the current state of things -- nothing prevents a miscompilation that generates `foo[0]` instead of `*foo`.

I suggest you add a `count-derefs` target (look at this for an example: https://github.com/FStarLang/karamel/blob/master/test/Makefile#L183) that counts the occurrences of `*` and makes sure it breaks if more or less dereferences are introduced in the generated C.

File: src/CStarToC11.ml [hunk 1]
```diff
@@ -921,14 +923,24 @@ and mk_stmts m stmts: C.stmt list =
...
+        Index (mk_expr m e1, c)
+    | e2' ->
+        Index (mk_expr m e1, e2')
+    end
 
 and mk_deref m (e: expr) : C.expr =
-  Deref (mk_expr m e)
+  match mk_expr m e with
+  | Address e' ->
+      (* *&expr is equivalent to expr *)
```
Comment (msprotz) [inline src/CStarToC11.ml:940]: according to the standard, "It is
always true that if E is a function designator or an lvalue that is a valid operand of the unary &
operator, *&E is a function designator or an lvalue equal to E." so looks like we're good here

**Review by msprotz — COMMENTED** [review]
Looks great thanks Tahina!

**Review by msprotz — APPROVED** [review]

**Comment by tahina-pro** [comment]
Thanks a lot Jonathan! I can now proceed with the F* companion PR.

**Pull request merged**